### ISSUE TITLE
PLANNING ISSUE: Spam comments moderation 

### ISSUE THREAD

**Issue by Gauravano** [publiclab/plots2#2304]
A user when banned by moderators, the alert text said that banning the author would remove all their content from the site -- but it did not. This is a bug.So, for handling this a comment moderation system needs to be built as discussed in #1785 .

* [x] make a "status" column for comments, default `1`, spam = `0`
* [x] make a "spam" method like this, that only moderators or admins can use (`user.role == admin`): https://github.com/publiclab/plots2/blob/master/app/models/node.rb#L158
* [x] a "publish" method which sets `status` back to 1
* [x] unit test the above 2 methods
* [x] display comments (on dashboard, /research, in answers, wiki pages and notes) only if they have `status == 1`
* [ ] display spammed comments to admins and moderators (`user.role == admin || user.role...`) but with text hidden, with a notice "A comment by @______ was put in moderation. _Read the full comment_" which displays the text, and a "republish" button

**Comment by Gauravano** [comment]
Hi, @jywarren as per the first part status column needs to be added but it's already there with the default value of 0 (for published). So, I am starting with 2nd part and giving spam comment as status = 2.

**Comment by jywarren** [comment]
This is looking awesome! Want to add "Planning issue" to the title? :-) 🎉

**Comment by Gauravano** [comment]
Hi @jywarren , sorry it was still open. Let's complete this ASAP.

**Comment by Gauravano** [comment]
Raised PR at #2957 . Also, I have gone through the work done by me in past. Just last 2 parts remaining.

The comments generated after PR https://github.com/publiclab/plots2/pull/2373 would be having `status = 1` but comments before that would be carrying `status = 0` . Could you please check stats @jywarren? 

We need to run a migration on production DB to change the status of all comments to 1 and the next step would be to merge #2957. Thanks!

**Comment by Gauravano** [comment]
> display spammed comments to admins and moderators (user.role == admin || user.role...) but with text hidden, with a notice "A comment by @______ was put in moderation. Read the full comment" which displays the text, and a "republish" button

@jywarren , any more details you wanna add to this part of issue?

**Comment by jywarren** [comment]
This is great. Maybe also that when a user is banned their comments should be moderated as well!

**Comment by Gauravano** [comment]
Adding this to checklist. @jywarren have you run migration for changing comment status to 1?

**Comment by jywarren** [comment]
Uhhhhh, hmm, sorry i think i need a refresher -- but any migrations that've been merged have definitely been run!

**Comment by Gauravano** [comment]
No issue, @jywarren refer this one https://github.com/publiclab/plots2/issues/2304#issuecomment-401536102

**Comment by Gauravano** [comment]
We ran a migration in #2373 but that is only for changing the default comment status to 1 so currently, production DB would be carrying comments with status 0 as well as 1. So, now we can run a query on production DB to change the status to 1. What do you think?

**Comment by jywarren** [comment]
That sounds good - let's do it as part of a migration though! You can
iterate through comments to do it in a migration, like in this example:
https://github.com/publiclab/plots2/blob/master/db/migrate/20130424190648_sync_ruser_id_drupal_users_uid.rb

**Comment by Gauravano** [comment]
Actually, in #2373 I included a statement for that but it didn't work that time, but I will try a PR with migration again. Thanks!

**Comment by jywarren** [comment]
ok awesome!

**Comment by jywarren** [comment]
Ah, oops reopening this as we still have a few items left like buttons. Thanks!

**Comment by Gauravano** [comment]
Hi @jywarren @steviepubliclab here's the progress on this issue:-
1. The comment moderation system is working, moderators can now mark any comment as spam
2. I just raised PR for the comments tab where all the spammed comments will appear
![spam_mod](https://user-images.githubusercontent.com/20878070/42710337-443cb958-8701-11e8-8700-d6d1d1ea2a2d.gif)

3. At start of the issue, one of the part was to add buttons on the profile page with each comment so that moderators can see if the spammy user has left any comments or not. The comment section looks like this right now 
![comment_section](https://user-images.githubusercontent.com/20878070/42710505-ecafd2e6-8701-11e8-9fcc-17d75a1195b5.png)

A bit loaded from data from my point of view. 
So, I want to know from you guys,  what do you think and where we can add `Mark as spam` button?
Thank you!

**Comment by Gauravano** [comment]
Hi @jywarren @steviepubliclab, we can also display spammy comments on the note only to moderator as previously decided as per checklist. What do you guys think?

 > display spammed comments to admins and moderators (user.role == admin || user.role...) but with text hidden, with a notice "A comment by @______ was put in moderation. Read the full comment" which displays the text, and a "republish" button

**Comment by steviepubliclab** [comment]
Hi!
Wow this is awesome!!! A couple questions:

1) What causes a comment to show up in the potential spam section? I
assuming first time commenters/posters?
2) Are first time comments now moderated until approved?
3) I'm not sure if you know this yet, or we've decided on it, but will the
moderation group get an email notification for new comments?
4) Can you tell me more about what you mean when you said "it's a bit
loaded from data?"

I think once spammy comments have been moderated, they don't need to show
up on notes to moderators. I think that answers your last question, but
clarify if I misread it.

This is soo great, thanks for working on this!
-Stevie

**Comment by jywarren** [comment]
Hi! I think we could probably make more columns in the table to better organize it so it's easier to skim through. What do you think?

**Comment by Gauravano** [comment]
Hi @steviepubliclab !
> 1) What causes a comment to show up in the potential spam section? I
assuming first-time commenters/posters?
If a moderator marks a comment as spam using "Mark as spam" button, then that comment would appear at /spam/comments 

> 2) Are first time comments now moderated until approved?
No, they are not, I can work on this one next. What do you say?

> 3) I'm not sure if you know this yet, or we've decided on it, but will the
moderation group gets an email notification for new comments?
Depends if we are implementing first-time comment moderation or not

> 4) Can you tell me more about what you mean when you said "it's a bit
loaded from data?"
Initially, we have planned to add "Mark as spam" button to each comment on the user profile page, but adding a button on current view can make it look messy. 
Agreed with @jywarren, we can organize it. 

Thanks!

**Comment by jywarren** [comment]
yes - this list is almost complete; but there are lots of small follow-ons and variations that could also help. Patrick suggested: `moderating a user would also moderate all their comments` and we can add that to this list.

The feature of holding all first-time comments for moderation would also be a bit of a project but we can add it to the list too if folks think it's time for that.

**Comment by gauravano** [comment]
Hi @jywarren, here's the new checklist for the follow-up parts of this issue.
[ ] Holding first-time comments for moderation 
[ ] Moderation of user affecting his comments 

@publiclab/community-reps please add more parts to this checklist and please give thumbs up :+1:  to those parts which you think are ready to be implemented. I would be happy to break each part of the checklist further down to small issues. Thanks!

**Comment by steviepubliclab** [comment]
Hi!  Sorry missed this one. Yes - I think we need to moderate first time comments. A lot of them have been some pretty bad spam lately! 

Thanks
-Stevie

**Comment by jywarren** [comment]
for ` Moderation of user affecting his comments` that means /all/ their comments are moderated when you moderate the user? 

I think this is good. But I think we may need to make the 'moderate comment' button NOT ban the user... or we need two buttons -- ban the user entirely AND all their comments, OR moderate just the one comment. 

This is a lot of extra work! I really appreciate it and I wonder if some of this could be made into tasks for GCI to reduce your workload at all?

**Comment by gauravano** [comment]
Yes, breaking both the parts into further parts is better. I would be starting with `Moderating first-timer comments` system. Thanks @jywarren @steviepubliclab !

**Comment by jywarren** [comment]
THANK YOU!!!
